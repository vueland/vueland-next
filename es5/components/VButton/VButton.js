"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VButton = void 0;

require("../../../src/components/VButton/VButton.scss");

var _vue = require("vue");

var _useColors2 = require("../../effects/use-colors");

var _useElevation2 = require("../../effects/use-elevation");

var _usePosition2 = require("../../effects/use-position");

var _VProgressCircular = require("../VProgressCircular");

var _helpers = require("../../helpers");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var VButton = (0, _vue.defineComponent)({
  name: 'v-button',
  props: _objectSpread({
    disabled: Boolean,
    outlined: Boolean,
    rounded: Boolean,
    round: Boolean,
    absolute: Boolean,
    loading: Boolean,
    left: Boolean,
    right: Boolean,
    text: Boolean,
    dark: Boolean,
    label: String,
    width: [String, Number],
    color: {
      type: String,
      "default": 'grey lighten-1'
    }
  }, (0, _useElevation2.elevationProps)()),
  emits: ['click'],
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        emit = _ref.emit;

    var _useColors = (0, _useColors2.useColors)(),
        setTextColor = _useColors.setTextColor,
        setBackground = _useColors.setBackground;

    var _useElevation = (0, _useElevation2.useElevation)(props),
        elevationClasses = _useElevation.elevationClasses;

    var _usePosition = (0, _usePosition2.usePosition)(props),
        positionClasses = _usePosition.positionClasses;

    var isFlat = (0, _vue.computed)(function () {
      return props.text || props.outlined;
    });
    var isLoadable = (0, _vue.computed)(function () {
      return !props.text && !props.outlined && props.loading;
    });
    var isElevetable = (0, _vue.computed)(function () {
      return !isLoadable.value && !props.disabled;
    });
    var classes = (0, _vue.computed)(function () {
      var elevations = isElevetable.value ? elevationClasses.value : {};
      return _objectSpread(_objectSpread({
        'v-button': true,
        'v-button--text': props.text || props.outlined,
        'v-button--outlined': props.outlined,
        'v-button--rounded': !props.round && props.rounded,
        'v-button--round': props.round,
        'v-button--disabled': props.disabled || isLoadable.value,
        'v-button--loading': props.loading
      }, elevations), positionClasses.value);
    });

    function genLoader() {
      return (0, _vue.h)('span', {
        "class": 'v-button__loader'
      }, slots.loading && slots.loading() || (0, _vue.h)(_VProgressCircular.VProgressCircular, {
        indeterminate: true,
        size: 23,
        width: 2
      }));
    }

    function genLabel() {
      var propsData = {
        "class": 'v-button__label'
      };
      return (0, _vue.h)('span', propsData, props.label);
    }

    function genContent() {
      return (0, _vue.h)('div', {
        "class": 'v-button__content'
      }, [props.label && genLabel(), slots["default"] && slots["default"]()]);
    }

    return function () {
      var setColor = isFlat.value ? setTextColor : setBackground;
      var width = props.width || 40;
      var propsData = {
        "class": classes.value,
        style: {
          width: (props.width || props.round) && (0, _helpers.convertToUnit)(width),
          height: props.round && (0, _helpers.convertToUnit)(width)
        },
        onClick: function onClick() {
          return !props.disabled && emit('click');
        }
      };
      return (0, _vue.h)('button', props.color && !isLoadable.value && !props.disabled ? setColor(props.color, propsData) : propsData, [genContent(), props.loading && genLoader()]);
    };
  }
});
exports.VButton = VButton;
//# sourceMappingURL=VButton.js.map