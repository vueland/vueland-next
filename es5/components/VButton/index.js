"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _VButton = require("./VButton");

Object.keys(_VButton).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (key in exports && exports[key] === _VButton[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _VButton[key];
    }
  });
});
//# sourceMappingURL=index.js.map