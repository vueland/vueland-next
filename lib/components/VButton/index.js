export * from "./VButton";
//# sourceMappingURL=index.js.map