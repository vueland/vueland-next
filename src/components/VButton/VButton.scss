@import "../../styles/scss/modules/global";

.v-button {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  position: relative;
  height: 40px;
  min-width: 40px;
  padding: 1px 6px;
  cursor: pointer;
  border-radius: 4px;
  overflow: hidden;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  color: map-get($white, 'base');
  font-size: .85rem;
  letter-spacing: .09rem;
  text-indent: .09rem;
  transition: .3s all $primary-transition;

  &:before {
    content: "";
    display: block;
    position: absolute;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: currentColor;
    opacity: 0;
    transition: all 0.3s ease;
  }

  &:hover {
    &:active {
      box-shadow: none !important;
    }
  }

  &--disabled {
    background: map-get($states, 'disabled');
    border-color: map-get($states, 'disabled');
    color: map-get($grey, 'darken-1');

    &.v-button--text {
      .v-button--loading {
        background: none !important;
      }
    }

    &.v-button--outlined {
      .v-button--loading {
        background: none !important;
        border-color: map-get($states, 'disabled');
      }
    }

    &:before {
      display: none;
    }
  }

  &--text {
    background: transparent;
    transition: all .2s;

    &:active {
      &:before {
        opacity: .4 !important;
      }
    }

    &:hover:before {
      opacity: .2;
    }
  }

  &--outlined {
    border: thin solid currentColor;

    & > * {
      color: inherit;
    }
  }

  &--rounded {
    border-radius: 50px;
  }

  &--round {
    border-radius: 50%;
  }

  &--loading {
    pointer-events: none;
    transition: none;

    .v-button__content {
      opacity: 0;
    }
  }

  &__content {
    display: flex;
    justify-content: center;
    align-items: center;
    flex: 1 0 auto;
    color: inherit;
    font-weight: 500;
  }

  &__loader {
    @include flex(flex, center, center);
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
  }

  &__label,
  &__loader {
    & > * {
      font-weight: 500;
    }
  }
}
