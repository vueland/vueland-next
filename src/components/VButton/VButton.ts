// Vue API
import { defineComponent, h, computed } from 'vue'

// Composable
import { useColors } from '../../composables/use-colors'
import { elevationProps, useElevation } from '../../composables/use-elevation'
import { usePosition } from '../../composables/use-position'

// Components
import { VProgressCircular } from '../VProgressCircular'

// Types
import { VNode } from 'vue'

// Helpers
import { convertToUnit } from '../../helpers'

export default defineComponent({
  name: 'v-button',
  props: {
    disabled: Boolean,
    outlined: Boolean,
    rounded: Boolean,
    round: Boolean,
    absolute: Boolean,
    loading: Boolean,
    left: Boolean,
    right: Boolean,
    text: Boolean,
    dark: Boolean,
    label: String,
    width: [String, Number],
    height: [String, Number],
    color: {
      type: String,
      default: 'grey lighten-1',
    },
    ...elevationProps(),
  } as any,
  emits: ['click'],

  setup(props, { slots, emit }): () => VNode {
    const {
      setTextClassNameColor,
      setBackgroundClassNameColor,
      setBackgroundCssColor,
      setTextCssColor,
    } = useColors()

    const { elevationClasses } = useElevation(props)
    const { positionClasses } = usePosition(props)

    const isFlat = computed<boolean>(() => {
      return props.text || props.outlined
    })

    const isLoadable = computed<boolean>(() => {
      return !props.text && !props.outlined && props.loading
    })

    const isElevetable = computed<boolean>(() => {
      return !isLoadable.value && !props.disabled
    })

    const classes = computed<Record<string, boolean>>(() => {
      const elevations = isElevetable.value ? elevationClasses.value : {}

      return {
        'v-button': true,
        'v-button--text': props.text || props.outlined,
        'v-button--outlined': props.outlined,
        'v-button--rounded': !props.round && props.rounded,
        'v-button--round': props.round,
        'v-button--disabled': props.disabled || isLoadable.value,
        'v-button--loading': props.loading,
        ...elevations,
        ...positionClasses.value,
        ...(!props.disabled && props.color && isFlat.value
          ? setTextClassNameColor(props.color)
          : {}),
        ...(!props.disabled && props.color && !isFlat.value
          ? setBackgroundClassNameColor(props.color)
          : {}),
      }
    })

    const styles = computed<Record<string, any>>(() => {
      const width = props.width || 40
      const height = props.round ? width: props.height ? props.height : ''

      return {
        width: (props.width || props.round) && convertToUnit(width),
        height: !!height && convertToUnit(height),
        ...(!props.disabled && props.color && isFlat.value ? setTextCssColor(props.color) : {}),
        ...(!props.disabled && props.color && !isFlat.value
          ? setBackgroundCssColor(props.color)
          : {}),
      }
    })

    const genLoader = (): VNode => {
      return h('span', {
          class: 'v-button__loader',
        },
        slots.loading?.() ||
        h(VProgressCircular, {
          indeterminate: true,
          size: 23,
          width: 2,
        }),
      )
    }

    const genLabel = (): VNode => {
      return h('span', {
        class: 'v-button__label',
      }, props.label)
    }

    const genContent = (): VNode => {
      return h('div', { class: 'v-button__content' },
        [(slots.default && slots.default()) || (props.label && genLabel())],
      )
    }

    return () =>
      h('button', {
          class: classes.value,
          style: styles.value,
          onClick: () => !props.disabled && emit('click'),
        },
        [genContent(), props.loading && genLoader()],
      )
  },
})
