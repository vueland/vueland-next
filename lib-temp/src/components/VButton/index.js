export * from './VButton';
import { VButton } from './VButton';
export default VButton;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuLi8uLi8uLi8uLi9zcmMvY29tcG9uZW50cy9WQnV0dG9uL2luZGV4LnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBLGNBQWMsV0FBVyxDQUFBO0FBRXpCLE9BQU8sRUFBRSxPQUFPLEVBQUUsTUFBTSxXQUFXLENBQUE7QUFFbkMsZUFBZSxPQUFPLENBQUEiLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnQgKiBmcm9tICcuL1ZCdXR0b24nXG5cbmltcG9ydCB7IFZCdXR0b24gfSBmcm9tICcuL1ZCdXR0b24nXG5cbmV4cG9ydCBkZWZhdWx0IFZCdXR0b25cbiJdfQ==