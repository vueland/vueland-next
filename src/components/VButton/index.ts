import VButton from './VButton'

export { VButton }
export default VButton
