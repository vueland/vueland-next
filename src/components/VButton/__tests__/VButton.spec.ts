import { mount } from "@vue/test-utils";
import { VButton } from "../VButton";
import "regenerator-runtime/runtime";

describe("VButton", () => {
  let mountFunction: (options?: any) => any;

  beforeEach(() => {
    mountFunction = (options = {}) => mount(VButton, { ...options });
  });

  it("should mount component and match snapshot", () => {
    const cmp = mountFunction();

    expect(cmp.html()).toMatchSnapshot();
  });

  it("should set color and match snapshot", () => {
    const cmp = mountFunction({
      props: {
        color: "red darken-3",
      },
    });

    expect(cmp.attributes().class).toContain("red darken-3");
    expect(cmp.html()).toMatchSnapshot();
  });

  it("should set button width and match snapshot", () => {
    const cmp = mountFunction({
      props: {
        width: 150,
      },
    });

    expect(cmp.attributes().style).toContain("min-width: 150px;");
    expect(cmp.html()).toMatchSnapshot();
  });

  it("should set label and match snapshot", () => {
    const cmp = mountFunction({
      props: {
        label: "test",
      },
    });

    expect(cmp.find(".v-button__label").text()).toContain("test");
    expect(cmp.html()).toMatchSnapshot();
  });

  it("should set disabled and match snapshot", async () => {
    const stub = jest.fn();
    const cmp = mountFunction({
      props: {
        disabled: true,
        onClick: stub,
      },
    });

    await cmp.trigger("click");

    expect(stub).toBeCalledTimes(0);
    expect(cmp.find(".v-button--disabled").exists()).toBe(true);
    expect(cmp.html()).toMatchSnapshot();
  });
});
