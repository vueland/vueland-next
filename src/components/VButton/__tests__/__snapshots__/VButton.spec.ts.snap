// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VButton should mount component and match snapshot 1`] = `
<button class="v-button">
</button>
`;

exports[`VButton should set color and match snapshot 1`] = `
<button class="v-button red darken-3">
</button>
`;

exports[`VButton should set disabled and match snapshot 1`] = `
<button class="v-button v-button--disabled">
</button>
`;

exports[`VButton should set label and match snapshot 1`] = `
<button class="v-button">
  <span class="v-button__label">
    test
  </span>
</button>
`;
